test/json_unittest.cc

```diff
@@ -0,0 +1,433 @@
+// (https://github.com/lloyd/yajl/issues/180)
+TEST(EdgeTest, StringWithNewline) {
+  GuardJsonException([](){
+    json::value v = json::Parser::FromString("\"\n\"");
+    EXPECT_EQ("\n", v->As<json::String>()->value());
+    // Output should be escaped with \n as that is more canonical.
+    EXPECT_TOSTRING_EQ("\"\n\"", v);
+  });
+}
+#endif
+
+TEST(EdgeTest, StringWithNonUnicodeEscape) {
+  GuardJsonException([](){
+    json::value v = json::Parser::FromString("\"foo\\nbar\"");
+    EXPECT_EQ("foo\nbar", v->As<json::String>()->value());
+    // Output should be escaped as that is more canonical.
+    EXPECT_TOSTRING_EQ("\"foo\\nbar\"", v);
+  });
+}
+
+TEST(EdgeTest, StringWithEveryEscape) {
+  GuardJsonException([](){
+    json::value v = json::Parser::FromString("\"x\\\\x\\/x\\bx\\fx\\nx\\rx\\tx\"");
+    EXPECT_EQ("x\\x/x\bx\fx\nx\rx\tx", v->As<json::String>()->value());
+    // Output should be escaped as that is more canonical.
+    // Since / does not need to be escaped, output does not escape that.
+    EXPECT_TOSTRING_EQ("\"x\\\\x/x\\bx\\fx\\nx\\rx\\tx\"", v);
+  });
+}
+
+TEST(EdgeTest, StringWithUnicodeEscape) {
+  GuardJsonException([](){
+    json::value v = json::Parser::FromString("\"foo\\u000abar\"");
+    EXPECT_EQ("foo\nbar", v->As<json::String>()->value());
+    // Output should be escaped as that is more canonical.
+    EXPECT_TOSTRING_EQ("\"foo\\nbar\"", v);
+  });
+}
+
+TEST(EdgeTest, StringWithUTF8) {
+  GuardJsonException([](){
+    // This is Korean for "Hello World!".
+    json::value v = json::Parser::FromString("\"안녕 세상아!\"");
+    EXPECT_EQ("안녕 세상아!", v->As<json::String>()->value());
+    EXPECT_TOSTRING_EQ("\"안녕 세상아!\"", v);
+  });
+}
+
+
+// Number edge cases
+
+TEST(EdgeTest, PositiveNumbers) {
+  GuardJsonException([](){
+    json::value v = json::Parser::FromString(
+      "[\n"
+      "  0, 0e+0, 0e-0, 0e0, 0E+0, 0E-0, 0E0,\n"
+      "  0.0, 0.0e+0, 0.0e-0, 0.0e0, 0.0E+0, 0.0E-0, 0.0E0,\n"
+      "  10, 10e+0, 10e-0, 10e0, 10E+0, 10E-0, 10E0,\n"
+      "  10.0, 10.0e+0, 10.0e-0, 10.0e0, 10.0E+0, 10.0E-0, 10.0E0,\n"
+      "  123, 123.456, 789, 1.05, 1.999e-99\n"
+      "]"
+    );
+    std::vector<double> numbers;
+    for (const auto& number : *v->As<json::Array>()) {
+      numbers.push_back(number->As<json::Number>()->value());
+    }
+    EXPECT_EQ(
+      std::vector<double>({
+        0, 0e+0, 0e-0, 0e0, 0E+0, 0E-0, 0E0,
+        0.0, 0.0e+0, 0.0e-0, 0.0e0, 0.0E+0, 0.0E-0, 0.0E0,
+        10, 10e+0, 10e-0, 10e0, 10E+0, 10E-0, 10E0,
+        10.0, 10.0e+0, 10.0e-0, 10.0e0, 10.0E+0, 10.0E-0, 10.0E0,
+        123, 123.456, 789, 1.05, 1.999e-99,
+      }),
+      numbers);
+    EXPECT_TOSTRING_EQ(
+      "["
+        "0.0,0.0,0.0,0.0,0.0,0.0,0.0,"
+        "0.0,0.0,0.0,0.0,0.0,0.0,0.0,"
+        "10.0,10.0,10.0,10.0,10.0,10.0,10.0,"
+        "10.0,10.0,10.0,10.0,10.0,10.0,10.0,"
+        "123.0,123.45600000000000307,789.0,"
+        "1.0500000000000000444,1.9989999999999998999e-99"
+      "]",
+      v
```

Comment:
Might as well close the parenthesis on this line. Also in other `EXPECT_TOSTRING_EQ` calls below.

Reply:
Done.